lib/ASTGen/Sources/ASTGen/LexicalLookup.swift

```diff
@@ -0,0 +1,611 @@
+  var isTheEndOfLookup: Bool {
+    flags.contains(.endOfLookup)
+  }
+
+  var shouldLookInMembers: Bool {
+    flags.contains(.shouldLookInMembers)
+  }
+
+  var resultPlacementRearranged: Bool {
+    flags.contains(.placementRearranged)
+  }
+
+  var shouldBeOmitted: Bool {
+    flags.contains(.shouldBeOmitted)
+  }
+
+  var shouldBeOptionallyOmitted: Bool {
+    flags.contains(.shouldBeOptionallyOmitted)
+  }
+
+  var ignoreNextFromHere: Bool {
+    flags.contains(.ignoreNextFromHere)
+  }
+
+  var lookupStopped: Bool {
+    flags.contains(.lookupStopped)
+  }
+
+  func consoleLogStr(sourceLocationConverter: SourceLocationConverter) -> String {
+    (isTheEndOfLookup ? "End here: " : "") + (resultPlacementRearranged ? "↕️ " : "")
+      + (ignoreNextFromHere ? "Ignore next from: " : "") + (shouldLookInMembers ? "Look memb: " : "\(name) ")
+      + sourceLocationConverter.location(for: position).lineWithColumn
+  }
+
+  static func == (lhs: ConsumedLookupResult, rhs: ConsumedLookupResult) -> Bool {
+    return lhs.rawName == rhs.rawName && lhs.position == rhs.position && lhs.flags == rhs.flags
+  }
+
+  func hash(into hasher: inout Hasher) {
+    hasher.combine(rawName)
+    hasher.combine(position)
+    hasher.combine(flags)
+  }
+}
+
+/// Determine behaviour during matching pass.
+struct ConsumedLookupResultFlag: OptionSet, Hashable {
+  let rawValue: Int
+
+  /// Indicates lookup ended at this name. Continue with
+  /// other names without matching and mark them as skipped.
+  static let endOfLookup = ConsumedLookupResultFlag(rawValue: 1 << 0)
+  /// This name prompts client to look in members of associated scope.
+  static let shouldLookInMembers = ConsumedLookupResultFlag(rawValue: 1 << 1)
+  /// The original position in result of this name
+  /// might be different than displayed.
+  static let placementRearranged = ConsumedLookupResultFlag(rawValue: 1 << 2)
+  /// The name should be ignored.
+  static let shouldBeOmitted = ConsumedLookupResultFlag(rawValue: 1 << 3)
+  /// If no match is found, this name should be ignored.
+  static let shouldBeOptionallyOmitted = ConsumedLookupResultFlag(rawValue: 1 << 4)
+  /// Means that one of the previous
+  /// names indicated the end of lookup.
+  static let lookupStopped = ConsumedLookupResultFlag(rawValue: 1 << 5)
+  /// Next names from associated position should be omitted.
+  /// Filtering is applied until then next name of this kind is found and
+  /// position used for ignoring is updated.
+  static let ignoreNextFromHere = ConsumedLookupResultFlag(rawValue: 1 << 6)
+}
+
+extension SourceLocation {
+  fileprivate var lineWithColumn: String {
+    return "\(line):\(column)"
+  }
+}
+
+extension String {
+  fileprivate func addPaddingUpTo(characters charCount: Int) -> String {
+    guard self.count < charCount else { return self }
+
+    let lengthDifference = charCount - self.count
+
+    var leftPad = ""
+    var rightPad = ""
+
+    for _ in 0..<Int(floor(Double(lengthDifference) / 2.0)) {
```

Comment:
Is this any different than `lengthDifference / 2`?

Reply:
Thanks for the suggestion, I've changed it to integer division.